src/lib.rs

```diff
@@ -17,244 +16,183 @@
 )]
 #![forbid(unsafe_code)]
 
-#[cfg(feature = "std")]
-#[macro_use]
-extern crate std;
+use ark_ff::{FpParameters, PrimeField};
+use ark_std::vec;
+use ark_std::vec::Vec;
+
+/// Infrastructure for the constraints counterparts.
+#[cfg(feature = "r1cs")]
+pub mod constraints;
 
-#[cfg(not(feature = "std"))]
-#[macro_use]
-extern crate alloc as std;
+mod absorb;
+pub use absorb::*;
 
-use ark_ff::models::{
-    Fp256, Fp256Parameters, Fp320, Fp320Parameters, Fp384, Fp384Parameters, Fp768, Fp768Parameters,
-    Fp832, Fp832Parameters,
-};
-use ark_ff::{to_bytes, PrimeField, ToConstraintField};
-use std::vec::Vec;
+/// The sponge for Poseidon
+///
+/// This implementation of Poseidon is entirely from Fractal's implementation in [COS20][cos]
+/// with small syntax changes.
+///
+/// [cos]: https://eprint.iacr.org/2019/1076
+pub mod poseidon;
 
 /// An enum for specifying the output field element size.
-#[derive(Clone)]
+#[derive(Clone, Copy, Eq, PartialEq)]
 pub enum FieldElementSize {
     /// Sample field elements from the entire field.
     Full,
 
-    /// Sample field elements from a subset of the field.
-    Truncated {
-        /// The maximum size of the subset is 2^num_bits.
-        num_bits: usize,
-    },
+    /// Sample field elements from a subset of the field, specified by the maximum number of bits.
+    Truncated(usize),
+}
+
+impl FieldElementSize {
+    pub(crate) fn num_bits<F: PrimeField>(&self) -> usize {
+        if let FieldElementSize::Truncated(num_bits) = self {
+            if *num_bits > (F::Params::MODULUS_BITS as usize) {
+                panic!("num_bits is greater than the capacity of the field.")
+            }
+            *num_bits.min(&(F::Params::CAPACITY as usize))
```

Comment:
We can probably remove min on this line now, since we panic if num_bits is greater.

Reply:
`CAPACITY` is `MODULUS_BITS - 1`. Removing this line would occasionally cause an overflow?

Reply:
Ah I missed that you now compare against MODULUS_BITS instead of CAPACITY in the check. It still seems strange that the size can equal MODULUS_BITS without panicking, but still gets bumped down without the user necessarily knowing (although it is a tiny bump).

@Pratyush: when you suggested comparing against MODULUS_BITS rather than CAPACITY, did we want to allow size to be able to be equal MODULUS_BITS? Otherwise, would it make more sense to directly compare against CAPACITY and remove the min?

Reply:
Where do we use this API? The reason I suggested `MODULUS_BITS` and not `CAPACITY` is two-fold: we want to eventually remove `CAPACITY` from the public API of `PrimeField` (for different reasons), and second, when I obtain a challenge of size `Full`, it contains `MODULUS_BITS` many bits, not `CAPACITY` many. (Because I'm converting an arbitrary field element to bits)

Reply:
just removed the `min`